### DISCUSSION TITLE
Render components using template directive

### DISCUSSION THREAD

**Discussion by ghost** [withastro/roadmap#783] [Proposal]
### Body

# Summary

This RFC proposes a way to render components using the template directive in Astro, enabling developers to insert component instances directly into their templates.

# Background & Motivation

In current Astro implementations, rendering components within templates can be challenging, as the `set:html={string}` directive injects an HTML string into an element, similar to setting `el.innerHTML`. This approach doesn't work for rendering Astro components or any framework components directly.

# Goals

- Enable developers to render Astro components or framework components using the template directive.
- Improve the developer experience by providing a more intuitive way to include components in templates.
- Ensure compatibility with the standard kebab-case format for all HTML attributes in Astro.

# Example

```astro
---
import Link from "../components/Link.astro";

const rawHTMLString = `Lorem ipsum <Link set:html='ipsum' href="website.com" />`;
---

<article>
  <h1>Render Components in Astro</h1>
  <p>This is an example of rendering an Astro component using the template directive:</p>
  <p set:html={rawHTMLString} />
</article>

**Comment by scriptogre** [comment]
I would also love to see this! 

Sometimes I want to use HTML as a prop, but it also includes a component.

```html
<Section
  title="Limited Spots"
  subtitle=`There are only <CountDown number={4} client:load /> spots left`
/>
```

In my opinion, the option above is mentally easier to follow than:

```html
<Section title="Limited Spots">
    <Fragment slot="subtitle">
        There are only <CountDown number={4} client:load /> spots left
    </Fragment>
</Section>
```

The second option feels a bit "heavier", but maybe that's just me.

I'm sure for most cases `<slots />` are the obvious choice, but for some cases, being able to use props with components instead would provide a really nice quality of life improvement in terms of readability. 

If anyone knows of an already existing better solution that I don't yet know of, please let me know!

**Reply by stereobooster** [reply]
In JSX this would look like:

```jsx
<Section
  title="Limited Spots"
  subtitle={<>There are only <CountDown number={4} client:load /> spots left</>}
/>
```
